python/ray/serve/metric.py

```diff
@@ -1,157 +1,178 @@
-                result_key = "{name}_{perc}th_perc_{window}_window".format(
-                    name=metric_name, perc=percentile, window=window)
-                aggregated_metric[result_key] = value
-
-        return aggregated_metric
-
-
-@ray.remote(num_cpus=0)
-def start_metric_monitor_loop(monitor_handle, duration_s=5):
-    while True:
-        ray.get(monitor_handle.scrape.remote())
-        time.sleep(duration_s)
+    async def push_batch(self, metadata, batch):
+        self._process_metadata(metadata)
+        self._process_batch(batch)
+
+    def _process_metadata(self, metadata):
+        for name, metadata in metadata.items():
+            if name not in self.metrics_cache:
+                constructor = metadata["type"].to_prometheus_class()
+                description = metadata["description"]
+                labels = metadata["labels"]
+
+                metric_object = constructor(
+                    name,
+                    description,
+                    labelnames=tuple(labels.keys()),
+                    registry=self.registry)
+                if labels:
+                    metric_object = metric_object.labels(**labels)
+                self.metrics_cache[name] = metric_object
+
+    def _process_batch(self, batch):
+        for event_type, name, value in batch:
+            assert name in self.metrics_cache, (
+                "Metrics {} was not registered.".format(name))
+            metric = self.metrics_cache[name]
+            if event_type == EventType.COUNTER:
+                metric.inc(value)
+            elif event_type == EventType.MEASURE:
+                metric.set(value)
+            else:
+                raise ValueError(
+                    "Unrecognized event type {}".format(event_type))
+
+
+@ray.remote
+class PrometheusSinkActor(PrometheusSink):
+    pass
+
+
+# The metric types are inspired by OpenTelemetry spec:
+# https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#three-kinds-of-instrument
+class Counter:
+    def __init__(self, collector, name):
+        self.collector = collector
+        self.name = name
+
+    def add(self, increment=1):
+        self.collector.push_event(EventType.COUNTER, self.name, increment)
+
+
+class Measure:
+    def __init__(self, collector, name):
+        self.collector = collector
+        self.name = name
+
+    def record(self, value):
+        self.collector.push_event(EventType.MEASURE, self.name, value)
+
+
+class EventType(enum.IntEnum):
+    COUNTER = 1
+    MEASURE = 2
+
+    def to_user_facing_class(self):
+        return {self.COUNTER: Counter, self.MEASURE: Measure}[self.value]
+
+    def to_prometheus_class(self):
+        return {self.COUNTER: PromCounter, self.MEASURE: PromGauge}[self.value]
+
+
+MetricMetadata = Dict[str, Dict]
+MetricEventBatch = List[Tuple[EventType, str, float]]
+
+
+class PushCollector:
+    def __init__(self,
+                 push_batch_callback: Callable[
```

Comment:
Is the `push_batch_callback` only used for testing right now? Fine if it is, but I think it would probably be simpler to pass in an object that inherits from some `MetricClient`